jumanji/environments/logic/sliding_tile_puzzle/env.py

```diff
@@ -0,0 +1,266 @@
+            reward_fn: RewardFn whose `__call__` method computes the reward of an environment
+                transition. The function must compute the reward based on the current state,
+                the chosen action and the next state.
+                Implemented options are [`DenseRewardFn`, `SparseRewardFn`,
+                `ImprovedDenseRewardFn`]. Defaults to `ImprovedDenseRewardFn`.
+            viewer: environment viewer for rendering.
+        """
+        self.generator = generator or RandomGenerator(grid_size=5)
+        self.reward_fn = reward_fn or ImprovedDenseRewardFn()
+
+        # Create viewer used for rendering
+        self._env_viewer = viewer or SlidingTilePuzzleViewer(name="SlidingTilePuzzle")
+        self.movements = jnp.array(
+            [[-1, 0], [1, 0], [0, -1], [0, 1]]  # Up  # Down  # Left  # Right
+        )
+        self.solved_puzzle = jnp.arange(self.generator.grid_size**2).reshape(
+            (self.generator.grid_size, self.generator.grid_size)
+        )
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[Observation]]:
+        """Resets the environment to an initial state."""
+        key, subkey = jax.random.split(key)
+        puzzle, empty_tile_position = self.generator(subkey)
+        state = State(
+            puzzle=puzzle,
+            empty_tile_position=empty_tile_position,
+            key=key,
+        )
+        action_mask = self._get_valid_actions(empty_tile_position)
+        obs = Observation(
+            puzzle=puzzle,
+            empty_tile_position=empty_tile_position,
+            action_mask=action_mask,
+        )
+        timestep = restart(observation=obs)
+        return state, timestep
+
+    def step(
+        self, state: State, action: chex.Array
+    ) -> Tuple[State, TimeStep[Observation]]:
+        """Updates the environment state after the agent takes an action."""
+        (updated_puzzle, updated_empty_tile_position) = self._move_empty_tile(
+            state.puzzle, state.empty_tile_position, action
+        )
+        # Check if the puzzle is solved
+        done = jnp.array_equal(updated_puzzle, self.solved_puzzle)
+
+        # Update the action mask
+        action_mask = self._get_valid_actions(updated_empty_tile_position)
+
+        next_state = State(
+            puzzle=updated_puzzle,
+            empty_tile_position=updated_empty_tile_position,
+            key=state.key,
+        )
+        obs = Observation(
+            puzzle=updated_puzzle,
+            empty_tile_position=updated_empty_tile_position,
+            action_mask=action_mask,
+        )
+
+        reward = self.reward_fn(state, action, next_state, self.solved_puzzle)
+
+        timestep = lax.cond(done, termination, transition, reward, obs)
+
+        return next_state, timestep
+
+    def _move_empty_tile(
+        self,
+        puzzle: chex.Array,
+        empty_tile_position: chex.Array,
+        action: chex.Array,
+    ) -> Tuple[chex.Array, chex.Array]:
```

Comment:
Because we're removing the cast below

```suggestion
    ) -> chex.Array:
```